File: finanze/infrastructure/repository/fetch/last_fetches_repository.py
from datetime import datetime
from uuid import UUID, uuid4

from application.ports.last_fetches_port import LastFetchesPort
from domain.entity import Feature
from domain.fetch_record import FetchRecord
from infrastructure.repository.db.client import DBClient
from infrastructure.repository.fetch.queries import LastFetchesQueries


def _map_row(row) -> FetchRecord:
    entity_account_id_raw = row["entity_account_id"]
    return FetchRecord(
        entity_id=UUID(row["entity_id"]),
        feature=Feature(row["feature"]),
        date=datetime.fromisoformat(row["date"]),
        entity_account_id=UUID(entity_account_id_raw)
        if entity_account_id_raw
        else None,
    )


class LastFetchesRepository(LastFetchesPort):
    def __init__(self, client: DBClient):
        self._db_client = client

    async def get_by_entity_id(self, entity_id: UUID) -> list[FetchRecord]:
        async with self._db_client.read() as cursor:
            await cursor.execute(
                LastFetchesQueries.GET_BY_ENTITY_ID,
                (str(entity_id),),
            )
            rows = await cursor.fetchall()
            return [_map_row(row) for row in rows]

    async def get_by_entity_account_id(
        self, entity_account_id: UUID
    ) -> list[FetchRecord]:
        async with self._db_client.read() as cursor:
            await cursor.execute(
                LastFetchesQueries.GET_BY_ENTITY_ACCOUNT_ID,
                (str(entity_account_id),),
            )
            rows = await cursor.fetchall()
            return [_map_row(row) for row in rows]

    async def save(self, fetch_records: list[FetchRecord]):
        async with self._db_client.tx() as cursor:
            for fetch_record in fetch_records:
                await cursor.execute(
                    LastFetchesQueries.UPSERT,
                    (
                        str(uuid4()),
                        str(fetch_record.entity_id),
                        fetch_record.feature.value,
                        fetch_record.date.isoformat(),
                        str(fetch_record.entity_account_id)
                        if fetch_record.entity_account_id
                        else None,
                    ),
                )

File: mobile/src/services/database/repositories/lastFetchesRepository.ts
import { DataManager } from "../dataManager"
import { Entity, Feature, FetchRecord } from "@/domain"
import { LastFetchesPort } from "@/application/ports"
import { LastFetchesQueries } from "./queries"

export class LastFetchesRepository implements LastFetchesPort {
  constructor(private client: DataManager) {}

  async getByEntityId(entityId: string): Promise<FetchRecord[]> {
    try {
      const result = await this.client.query<any>(
        LastFetchesQueries.GET_BY_ENTITY_ID,
        [entityId],
      )
      return result.rows.map(row => ({
        entityId: row.entity_id,
        feature: row.feature as Feature,
        date: row.date,
      }))
    } catch {
      return []
    }
  }

  async getAll(): Promise<FetchRecord[]> {
    try {
      const result = await this.client.query<any>(
        "SELECT entity_id, feature, date FROM last_fetches",
      )
      return result.rows.map(row => ({
        entityId: row.entity_id,
        feature: row.feature as Feature,
        date: row.date,
      }))
    } catch {
      return []
    }
  }

  async getGroupedByEntity(
    feature: Feature,
  ): Promise<Map<Entity, FetchRecord>> {
    try {
      const result = await this.client.query<any>(
        LastFetchesQueries.GET_GROUPED_BY_ENTITY,
        [feature],
      )

      const map = new Map<Entity, FetchRecord>()
      for (const row of result.rows) {
        const entity: Entity = {
          id: row.entity_id,
          name: row.entity_name,
          naturalId: row.entity_natural_id ?? null,
          type: row.entity_type,
          origin: row.entity_origin,
          iconUrl: row.icon_url,
        }
        const record: FetchRecord = {
          entityId: row.entity_id,
          feature: row.feature as Feature,
          date: row.date,
        }
        map.set(entity, record)
      }
      return map
    } catch {
      return new Map()
    }
  }

  async save(fetchRecords: FetchRecord[]): Promise<void> {
    for (const record of fetchRecords) {
      await this.client.execute(LastFetchesQueries.UPSERT, [
        record.entityId,
        record.feature,
        record.date,
      ])
    }
  }
}
